src/browser-polyfill.js

```diff
@@ -6,7 +6,48 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 "use strict";
 
-if (typeof browser === "undefined") {
+/**
+ * Returns true if the given object is an object with a `then` method, and can
+ * therefore be assumed to behave as a Promise.
+ *
+ * @param {*} value The value to test.
+ * @returns {boolean} True if the value is thenable.
+ */
+const isThenable = value => {
+  return value && typeof value === "object" && typeof value.then === "function";
+};
+
+/**
+ * Checks if the current browser supports
+ * the Promise-based WebExtension APIs.
+ *
+ * @returns {boolean} True if yes.
+ */
+const supportsPromises = () => {
+  if (typeof browser === "undefined") {
+    return false;
+  }
+  // If `browser.runtime.lastError` doesn’t exist, assume promises are supported.
+  if (browser.runtime && typeof browser.runtime.lastError === "undefined") {
+    return true;
+  }
+  if (browser.extension && typeof browser.extension.lastError === "undefined") {
+    return true;
+  }
+  try {
+    return isThenable(browser.runtime.getPlatformInfo());
+  } catch (error) {
+    // Microsoft Edge doesn’t support `browser.runtime.getPlatformInfo()`
+    try {
```

Comment:
There’s not need to nest this. If there are no errors, the function will return before the second try/catch